File: reef-tang/tang/src/main/java/org/apache/reef/tang/formats/ParameterParser.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.reef.tang.formats;

import org.apache.reef.tang.ExternalConstructor;
import org.apache.reef.tang.exceptions.BindException;
import org.apache.reef.tang.util.MonotonicTreeMap;
import org.apache.reef.tang.util.ReflectionUtilities;

import java.lang.reflect.Constructor;
import java.lang.reflect.Type;
import java.util.Collections;
import java.util.HashSet;
import java.util.Set;

public class ParameterParser {
  private static final Set<String> BUILTIN_NAMES = new HashSet<String>() {
    private static final long serialVersionUID = 1L;

    {
      Collections.addAll(this,
          String.class.getName(),
          Byte.class.getName(),
          Character.class.getName(),
          Short.class.getName(),
          Integer.class.getName(),
          Long.class.getName(),
          Float.class.getName(),
          Double.class.getName(),
          Boolean.class.getName(),
          Void.class.getName());
    }
  };
  MonotonicTreeMap<String, Constructor<? extends ExternalConstructor<?>>> parsers = new MonotonicTreeMap<>();

  @SuppressWarnings({"unchecked", "rawtypes"})
  public void addParser(Class<? extends ExternalConstructor<?>> ec) throws BindException {
    Class<?> tc = (Class<?>) ReflectionUtilities.getInterfaceTarget(
        ExternalConstructor.class, ec);
    addParser((Class) tc, (Class) ec);
  }

  public <T, U extends T> void addParser(Class<U> clazz, Class<? extends ExternalConstructor<T>> ec) throws BindException {
    Constructor<? extends ExternalConstructor<T>> c;
    try {
      c = ec.getDeclaredConstructor(String.class);
      c.setAccessible(true);
    } catch (NoSuchMethodException e) {
      throw new BindException("Constructor "
          + ReflectionUtilities.getFullName(ec) + "(String) does not exist!", e);
    }
    c.setAccessible(true);
    parsers.put(ReflectionUtilities.getFullName(clazz), c);
  }

  public void mergeIn(ParameterParser p) {
    for (String s : p.parsers.keySet()) {
      if (!parsers.containsKey(s)) {
        parsers.put(s, p.parsers.get(s));
      } else {
        if (!parsers.get(s).equals(p.parsers.get(s))) {
          throw new IllegalArgumentException(
              "Conflict detected when merging parameter parsers! To parse " + s
                  + " I have a: " + ReflectionUtilities.getFullName(parsers.get(s).getDeclaringClass())
                  + " the other instance has a: " + ReflectionUtilities.getFullName(p.parsers.get(s).getDeclaringClass()));
        }
      }
    }
  }

  public <T> T parse(Class<T> c, String s) {
    Class<?> d = ReflectionUtilities.boxClass(c);
    for (Type e : ReflectionUtilities.classAndAncestors(d)) {
      String name = ReflectionUtilities.getFullName(e);
      if (parsers.containsKey(name)) {
        T ret = parse(name, s);
        if (c.isAssignableFrom(ret.getClass())) {
          return ret;
        } else {
          throw new ClassCastException("Cannot cast from " + ret.getClass() + " to " + c);
        }
      }
    }
    return parse(ReflectionUtilities.getFullName(d), s);
  }

  @SuppressWarnings("unchecked")
  public <T> T parse(String name, String value) {
    if (parsers.containsKey(name)) {
      try {
        return (T) (parsers.get(name).newInstance(value).newInstance());
      } catch (ReflectiveOperationException e) {
        throw new IllegalArgumentException("Error invoking constructor for "
            + name, e);
      }
    } else {
      if (name.equals(String.class.getName())) {
        return (T) value;
      }
      if (name.equals(Byte.class.getName())) {
        return (T) (Byte) Byte.parseByte(value);
      }
      if (name.equals(Character.class.getName())) {
        return (T) (Character) value.charAt(0);
      }
      if (name.equals(Short.class.getName())) {
        return (T) (Short) Short.parseShort(value);
      }
      if (name.equals(Integer.class.getName())) {
        return (T) (Integer) Integer.parseInt(value);
      }
      if (name.equals(Long.class.getName())) {
        return (T) (Long) Long.parseLong(value);
      }
      if (name.equals(Float.class.getName())) {
        return (T) (Float) Float.parseFloat(value);
      }
      if (name.equals(Double.class.getName())) {
        return (T) (Double) Double.parseDouble(value);
      }
      if (name.equals(Boolean.class.getName())) {
        return (T) (Boolean) Boolean.parseBoolean(value);
      }
      if (name.equals(Void.class.getName())) {
        throw new ClassCastException("Can't instantiate void");
      }
      throw new UnsupportedOperationException("Don't know how to parse a " + name);
    }
  }

  public boolean canParse(String name) {
    return parsers.containsKey(name) || BUILTIN_NAMES.contains(name);
  }
}

File: lang/cs/Org.Apache.REEF.Tang/Implementations/ClassHierarchy/ParameterParser.cs
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

using System;
using System.Collections.Generic;
using System.Reflection;
using Org.Apache.REEF.Tang.Exceptions;
using Org.Apache.REEF.Tang.Interface;
using Org.Apache.REEF.Tang.Util;
using Org.Apache.REEF.Utilities.Logging;

namespace Org.Apache.REEF.Tang.Implementations.ClassHierarchy
{
    public class ParameterParser
    {
        private static readonly Logger LOGGER = Logger.GetLogger(typeof(ParameterParser));

        readonly MonotonicTreeMap<String, ConstructorInfo> parsers = new MonotonicTreeMap<String, ConstructorInfo>();

        // ec: ACons, tc: A
        public void AddParser(Type ec)
        {
            Type tc = (Type)ReflectionUtilities.GetInterfaceTarget(typeof(IExternalConstructor<>), ec);
            AddParser(tc, ec);
        }

        // TODO
        // public  <T, U extends T> void AddParser(Class<U> clazz, Class<? extends ExternalConstructor<T>> ec) throws BindException {
        // public void AddParser<T, U, V>(GenericType<U> clazz, GenericType<V> ec) 
        //    where U : T
        //    where V:  IExternalConstructor<T>
        // {

        // clazz: A, ec: ACons
        private void AddParser(Type clazz, Type ec)
        {
            ConstructorInfo c = ec.GetConstructor(BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Instance, null, new Type[] { typeof(string) }, null);
            
            if (c == null)
            {
                var e = new BindException("Constructor " + ReflectionUtilities.GetAssemblyQualifiedName(ec) + "(String) does not exist!");
                Org.Apache.REEF.Utilities.Diagnostics.Exceptions.Throw(e, LOGGER);
            }

            // c.setAccessible(true); // set as public 
            parsers.Add(ReflectionUtilities.GetAssemblyQualifiedName(clazz), c);
        }


        public void MergeIn(ParameterParser p)
        {
            foreach (string s in p.parsers.Keys)
            {
                if (!parsers.ContainsKey(s))
                {
                    ConstructorInfo ci;
                    p.parsers.TryGetValue(s, out ci);
                    parsers.Add(s, ci);
                }
                else
                {
                    ConstructorInfo oldC;
                    ConstructorInfo newC;
                    parsers.TryGetValue(s, out oldC);
                    p.parsers.TryGetValue(s, out newC);
                    if (!oldC.Equals(newC))
                    {
                        var e = new ArgumentException(
                        "Conflict detected when merging parameter parsers! To parse " + s
                        + " I have a: " + ReflectionUtilities.GetAssemblyQualifiedName(oldC.DeclaringType)
                        + " the other instance has a: " + ReflectionUtilities.GetAssemblyQualifiedName(newC.DeclaringType));
                        Org.Apache.REEF.Utilities.Diagnostics.Exceptions.Throw(e, LOGGER);
                    }
                }
            }
        }

        // (Integer, "3") return object of new Integer(3)
        public object Parse(Type c, String s)
        {
            Type d = ReflectionUtilities.BoxClass(c);
            foreach (Type e in ReflectionUtilities.ClassAndAncestors(d)) // get all the super classes of Integer for example
            {
                string name = ReflectionUtilities.GetAssemblyQualifiedName(e);
                if (parsers.ContainsKey(name))
                {
                    object ret = Parse(name, s);
                    if (c.IsAssignableFrom(ret.GetType())) // check if ret can be cast as c
                    {
                        return ret;
                    }
                    else
                    {
                        var ex = new InvalidCastException("Cannot cast from " + ret.GetType() + " to " + c);
                        Org.Apache.REEF.Utilities.Diagnostics.Exceptions.Throw(ex, LOGGER);
                    }
                }
            }
            return Parse(d.Name, s);
        }

        // name: "Integer", value: "12"
        public object Parse(string name, string value)
        {
            if (parsers.ContainsKey(name))
            {
                try
                {
                    ConstructorInfo c = null;
                    parsers.TryGetValue(name, out c);
                    var o = c.Invoke(new object[] { value });
                    var m = o.GetType().GetMethod("NewInstance", BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance);
                    return m.Invoke(o, new object[] { });
                }
                catch (TargetInvocationException e)
                {
                    Org.Apache.REEF.Utilities.Diagnostics.Exceptions.Caught(e, Level.Error, LOGGER);
                    Org.Apache.REEF.Utilities.Diagnostics.Exceptions.Throw(new ArgumentException("Error invoking constructor for " + name, e), LOGGER);
                }
            }
            else if (name.Equals(typeof(string).Name))
            {
                return (object)value;
            }
            if (name.Equals(typeof(Byte).Name))
            {
                return (object)(Byte)Byte.Parse(value);
            }
            if (name.Equals(typeof(Char).Name))
            {
                return (object)(Char)value[0];
            }
            if (name.Equals(typeof(short).Name))
            {
                return (System.Int16)System.Int16.Parse(value);
            }
            if (name.Equals(typeof(int).Name))
            {
                return (object)(Int32)Int32.Parse(value);
            }
            if (name.Equals(typeof(long).Name))
            {
                return (object)(Int64)Int64.Parse(value);
            }
            if (name.Equals(typeof(float).Name))
            {
                return (object)(Single)Single.Parse(value);
            }
            if (name.Equals(typeof(Double).Name))
            {
                return (object)(Double)Double.Parse(value);
            }
            if (name.Equals(typeof(Boolean).Name))
            {
                return (object)(Boolean)Boolean.Parse(value);
            }
            if (name.Equals(typeof(byte[]).Name))  
            {
                byte[] bytes = new byte[value.Length * sizeof(char)];
                System.Buffer.BlockCopy(value.ToCharArray(), 0, bytes, 0, bytes.Length);
                return bytes;
            }
            return null; 
        }

        private static readonly ISet<string> BUILTIN_NAMES = new HashSet<string>(new string[]
            {
                typeof(string).AssemblyQualifiedName, 
                typeof(Byte).AssemblyQualifiedName,
                typeof(char).AssemblyQualifiedName,
                typeof(short).AssemblyQualifiedName,
                typeof(Int32).AssemblyQualifiedName,
                typeof(long).AssemblyQualifiedName,
                typeof(float).AssemblyQualifiedName,
                typeof(double).AssemblyQualifiedName,
                typeof(bool).AssemblyQualifiedName
            }); 
        
        public bool CanParse(string name)
        {
            return parsers.ContainsKey(name) || BUILTIN_NAMES.Contains(name);
        }
    }
}
